flask/cli.py

```diff
@@ -33,6 +33,14 @@
     dotenv = None
 
 
+def _called_with_wrong_args(factory, exc_info):
+    exc_type, exc_value, tb = exc_info
+    return exc_type is TypeError and \
+        str(exc_value).startswith((
+            '%s() takes' % factory.__name__,
+            '%s() missing' % factory.__name__))
+
```

Comment:
@mitsuhiko are there situations that will raise a `TypeError` that has `factory.__name__` in it but isn't actually related to calling the factory? Might be able to simplify the check instead of trying to cover a bunch of different messages.

Reply:
Good inspiration!
```python
while tb.tb_next:
    tb = tb.tb_next
return tb.tb_frame.f_code.co_name != factory.__name__
```

Reply:
Mostly good, just need a bit of rearranging to account for a recursive call failing.

```python
def create_app():
    create_app('bad')
```

Reply:
```python
def other():
    raise TypeError

def create_app():
    other()
```

prints the wrong message with that, need to adjust it to stop once we hit the factory in the traceback.